feat: use nivo responsive bar chart for progress and criticality bar

meretp opened the pull request:
### Summary of changes

To improve the progress bar this PR adds a new library nivo and introduces a ResponsiveBar. 

### Context and reason for change

fixes part of #2622 

### How can the changes be tested

Open OpossumUI and play around with the progress/criticality bar.

Note: Please review the [guidelines for contributing](https://github.com/opossum-tool/OpossumUI/blob/main/CONTRIBUTING.md) to this repository.

meretp commented on src/Frontend/Components/ProgressBar/__tests__/ProgressBar.test.tsx:
I decided to replace this unit test with an e2e test.

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+    fontWeight: 500,
+    position: 'relative',
+    top: 100,
+    'z-index': 100,
```
this is not a valid css prop as is and should be `zIndex`. but given that it doesn't do anything right now we should probably just delete it then.

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+    <MuiBox aria-label={'tooltip'} style={classes.tooltip}>
+      Number of resources...
+      <div style={classes.div}>
+        <span
+          style={{
+            ...classes.colorBlock,
+            background: OpossumColors.pastelDarkGreen,
+          }}
+        ></span>
```
here and everywhere else:

```suggestion
        <span
          style={{
            ...classes.colorBlock,
            background: OpossumColors.pastelDarkGreen,
          }}
        />
```

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -70,129 +138,47 @@ export function getCriticalityBarTooltipText(
+          …containing highly critical signals:{' '}
+          {new Intl.NumberFormat().format(
+            progressBarData.filesWithHighlyCriticalExternalAttributionsCount,
+          )}{' '}
```
here and below: i don't see why we need a trailing space after the item

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+    margin: 2,
+    fontWeight: 500,
+    position: 'relative',
+    top: 100,
```
the positioning is not working consistently: depending on the bar's mode, the tooltip appears higher or lower below the cursor. could you find a way to make the positioning consistent?

meretp replied:
This is still to do, but so far I couldn't find a way.

mstykow replied:
ok, i'll have a look

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+    color: 'white',
+    fontFamily: 'Karla Variable, sans-serif',
+    padding: 4,
+    fontSize: 10,
```
to me this seems quite small. i'm not aware of us using 10px anywhere else. i've seen 12px for captions and i think that's the smallest size we should use.

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+    top: 100,
+    'z-index': 100,
+  } as CSSProperties,
+  div: { display: 'flex', 'align-items': 'center' },
```
the name "div" for a css class that's applied to any of the many divs an app may have is really not the best. it's better to choose class names by the function they fulfill.

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.tsx:
```diff
@@ -49,31 +43,78 @@ export function ProgressBar(props: ProgressBarProps): ReactElement {
+      <ResponsiveBar
+        data={data}
+        keys={Object.keys(data[0])}
+        margin={{ top: 5, right: 1, bottom: 5, left: 6 }}
```
these margins are asymmetrical and deviate from the multiples-of-4 rule. i think there is no reason to break the rules and introduce asymmetry here:
```suggestion
        margin={{ top: 4, bottom: 4 }}
```

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.tsx:
```diff
@@ -49,31 +43,78 @@ export function ProgressBar(props: ProgressBarProps): ReactElement {
+  const data: Array<{ [key: string]: number }> = props.showCriticalSignals
+    ? [
+        {
+          high_critical:
```
here and below: JavaScript uses camel case

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.tsx:
```diff
@@ -4,28 +4,22 @@
     '&:hover': {
       cursor: 'pointer',
-      opacity: 0.75,
+      opacity: 1,
```
do we still need this opacity declaration?

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.tsx:
```diff
@@ -49,31 +43,78 @@ export function ProgressBar(props: ProgressBarProps): ReactElement {
       ? resourcesWithCriticalExternalAttributions
       : props.progressBarData.resourcesWithNonInheritedExternalAttributionOnly,
   );
+  const data: Array<{ [key: string]: number }> = props.showCriticalSignals
```
could you use nivo's built-in type (`BarDatum`) to ensure that the object we're defining here always matches what nivo expects, even if nivo gets refactored?

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
 
+const classes = {
+  tooltip: {
+    flex: 1,
```
this prop doesn't seem to have any effect

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,95 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
 }
 
+const classes = {
+  tooltip: {
```
could you give the interior of the tooltip a little more room to breathe? right now the lines of text have hardly any space between them and are also really close to the edge of the container.

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,93 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+    fontSize: 12,
+    margin: 2,
+    fontWeight: 500,
+    // position: 'relative',
+    // top: 'auto',
```
can we delete these commented props?

meretp replied:
sure, I was trying to debug and find a solution for the positioning, I reverted it to the previous state.

mstykow commented on src/Frontend/Components/ProgressBar/ProgressBar.util.tsx:
```diff
@@ -37,27 +37,93 @@ export function useOnProgressBarClick(resourceIds: Array<string>) {
+          }}
+        />
+        <span>
+          ...with attributions:{' '}
```
here and above: let's use either three separate dots "..." or the special three dot symbol "…" consistently.

meretp replied:
done

mstykow commented:
The following issues need to be solved before this PR can merge:
1. the position differs between tooltip modes
2. when resizing the window down, the new bar chart does not resize itself thus causing x-overflow

meretp commented:
Added some commits that fixes the second issue and updates nivo bar - there has been a new release in the meantime. Also I updated the diff to use the ariaLabel from ResponsiveBar.

meretp commented:
To fix the first issue the only idea I could come up with is to use sepearate CSS styling like
 ```
transformProgress: { transform: 'translate(0px, 168px)' } as CSSProperties,
 transformCritical: { transform: 'translate(0px, 144px)' } as CSSProperties,
```
to transform the tooltips separately. This does not feel like a stable solution so I only add this as a comment not commit yet, what do you think?

meretp commented:
Have to take a look at why the unit tests are failing now

mstykow commented:
Closing for now since no work done in a while.
